ledger-service/http-json/src/main/scala/com/digitalasset/http/json/JsonProtocol.scala

```diff
@@ -238,33 +239,58 @@ object JsonProtocol extends DefaultJsonProtocol with ExtraFormats {
     *  1. template IDs are required
     *  2. query key may be absent
     *  3. special error if you appear to have queried outside 'query'
+    *
+    *  This provides an (almost) consistent behavior when reading the 'templateIds' and
+    *  'query' fields. Further extra fields may be added by concrete implementations.
     */
-  implicit val GetActiveContractsRequestFormat: RootJsonReader[domain.GetActiveContractsRequest] = {
-    case class GACR(
+  private def requestJsonReader[Request](validExtraFields: Set[String])(
+      toRequest: (
+          OneAnd[Set, TemplateId.OptionalPkg],
+          Map[String, JsValue],
+          Map[String, JsValue],
+      ) => Request
+  ): RootJsonReader[Request] = {
+    final case class BaseRequest(
         templateIds: Set[domain.TemplateId.OptionalPkg],
         query: Option[Map[String, JsValue]],
     )
-    val validKeys = Set("templateIds", "query")
-    implicit val primitive: JsonReader[GACR] = jsonFormat2(GACR.apply)
+    val validKeys = Set("templateIds", "query") ++ validExtraFields
+    implicit val primitive: JsonReader[BaseRequest] = jsonFormat2(BaseRequest.apply)
     jsv => {
-      val GACR(tids, q) = jsv.convertTo[GACR]
-      val extras = jsv.asJsObject.fields.keySet diff validKeys
-      if (extras.nonEmpty)
+      val BaseRequest(tids, query) = jsv.convertTo[BaseRequest]
+      val unsupported = jsv.asJsObject.fields.keySet diff validKeys
+      if (unsupported.nonEmpty)
         deserializationError(
-          s"unsupported query fields ${extras}; likely should be within 'query' subobject"
+          s"unsupported query fields $unsupported; likely should be within 'query' subobject"
         )
-      tids.headOption.cata(
-        h => domain.GetActiveContractsRequest(OneAnd(h, tids - h), q getOrElse Map.empty),
+      val extraFields = jsv.asJsObject.fields.filter { case (fieldName, _) =>
+        validExtraFields(fieldName)
+      }
+      val nonEmptyTids = tids.headOption.cata(
+        h => OneAnd(h, tids - h),
         deserializationError("search requires at least one item in 'templateIds'"),
       )
+      toRequest(nonEmptyTids, query.getOrElse(Map.empty), extraFields)
     }
   }
 
+  implicit val GetActiveContractsRequestFormat: RootJsonReader[domain.GetActiveContractsRequest] =
+    requestJsonReader(validExtraFields = Set.empty)((tids, query, _) =>
+      domain.GetActiveContractsRequest(tids, query)
+    )
+
+  implicit val SearchForeverQueryFormat: RootJsonReader[domain.SearchForeverQuery] = {
+    val OffsetKey = "queryOffset"
```

Comment:
Given that once we set this, we’re stuck with this ~forever~ a very long time could you expand on your earlier comment that calling this `offset` results in messages being reinterpreted? I don’t understand why we cannot disambiguate between the two requests.

Reply:
This used to make some earlier tests fail. I thought about it a bit and what the current code does is just check for the presence of `offset` in the object root and reject this with a mistake if an `offset` message was passed already. Thinking better about it, the per-query `offset`s are always in the array that specifies the queries, not in the root, so I have no idea why this failed initially. Probably something else was broken at that point. Changing this to `offset`.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/9847/commits/7575a67b643248a31d037dc60fe9d8d65539bb44